examples/batched-solver/batched-solver.cpp

```diff
@@ -108,46 +147,205 @@ int main(int argc, char *argv[])
+    // std::uniform_real_distribution<real_type> distb(0.75, 1.0);
+    std::normal_distribution<real_type> distb(0.5, 0.1);
+    std::vector<real_type> spacings(nsystems * nrows);
+    std::generate(spacings.begin(), spacings.end(),
+                  [&]() { return distb(rgen); });
+
+    std::vector<value_type> allvalues(nnz * nsystems);
+    for (size_type isys = 0; isys < nsystems; isys++) {
+        allvalues[isys * nnz] = 2.0 / spacings[isys * nrows];
+        allvalues[isys * nnz + 1] = -1.0;
+        for (int irow = 0; irow < nrows - 2; irow++) {
+            allvalues[isys * nnz + 2 + irow * 3] = -1.0;
+            allvalues[isys * nnz + 2 + irow * 3 + 1] =
+                2.0 / spacings[isys * nrows + irow + 1];
+            allvalues[isys * nnz + 2 + irow * 3 + 2] = -1.0;
+        }
+        allvalues[isys * nnz + 2 + (nrows - 2) * 3] = -1.0;
+        allvalues[isys * nnz + 2 + (nrows - 2) * 3 + 1] =
+            2.0 / spacings[(isys + 1) * nrows - 1];
+        assert(isys * nnz + 2 + (nrows - 2) * 3 + 2 == (isys + 1) * nnz);
+    }
+
+    std::vector<index_type> rowptrs(nrows + 1);
+    rowptrs[0] = 0;
+    rowptrs[1] = 2;
+    for (int i = 2; i < nrows; i++) {
+        rowptrs[i] = rowptrs[i - 1] + 3;
+    }
+    rowptrs[nrows] = rowptrs[nrows - 1] + 2;
+    assert(rowptrs[nrows] == nnz);
+
+    std::vector<index_type> colidxs(nnz);
+    colidxs[0] = 0;
+    colidxs[1] = 1;
+    const int nnz_per_row = 3;
+    for (int irow = 1; irow < nrows - 1; irow++) {
+        colidxs[2 + (irow - 1) * nnz_per_row] = irow - 1;
+        colidxs[2 + (irow - 1) * nnz_per_row + 1] = irow;
+        colidxs[2 + (irow - 1) * nnz_per_row + 2] = irow + 1;
+    }
+    colidxs[2 + (nrows - 2) * nnz_per_row] = nrows - 2;
+    colidxs[2 + (nrows - 2) * nnz_per_row + 1] = nrows - 1;
+    assert(2 + (nrows - 2) * nnz_per_row + 1 == nnz - 1);
+
+    std::vector<value_type> allb(nrows * nsystems);
+    for (size_type isys = 0; isys < nsystems; isys++) {
+        const value_type bval = distb(rgen);
+        std::fill(allb.begin() + isys * nrows,
+                  allb.begin() + (isys + 1) * nrows, bval);
+    }
+
+    index_type* const row_ptrs = exec->alloc<index_type>(nrows + 1);
+    exec->copy_from(exec->get_master().get(), static_cast<size_type>(nrows + 1),
+                    rowptrs.data(), row_ptrs);
+    index_type* const col_idxs = exec->alloc<index_type>(nnz);
+    exec->copy_from(exec->get_master().get(), static_cast<size_type>(nnz),
+                    colidxs.data(), col_idxs);
+    value_type* const all_values = exec->alloc<value_type>(nsystems * nnz);
+    exec->copy_from(exec->get_master().get(), nsystems * nnz, allvalues.data(),
+                    all_values);
+    value_type* const all_b = exec->alloc<value_type>(nsystems * nrows);
+    exec->copy_from(exec->get_master().get(), nsystems * nrows, allb.data(),
+                    all_b);
+    return {nsystems, nrows, nnz, row_ptrs, col_idxs, all_values, all_b};
```

Comment:
Is there any reason not to use array?
it requires the manually free afterward. Also the clean up does not need it actually

Reply:
Here the idea is to show integration with applications. It's not unusual for applications to directly deal with pointers. Even if not, that is the lowest common denominator for communication between the application and Ginkgo. So I thought it would be a good idea to show interfacing using raw pointers.

Reply:
I see, then maybe wrap a destructor of the structure should be better than using array

Reply:
I wanted to keep the struct and example simple... if I add a destructor, then it makes sense to add a constructor, and it'll start to get quite complex. What I had in mind was that if you have a C-like code, and you're willing to compile some files using C++, you can use GInkgo.

Reply:
the constructor and destructor are like what you did, I think.